File: CrossDroid.Windows/Backend/Network/CrossDroidProtocol.cs
using System;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace CrossDroid.Windows.Backend.Network;

public enum MessageType
{
    PairRequest,
    PairResponse,
    TransferOffer,
    TransferAccept,
    TransferReject,
    TransferCancel,
    FileChunk
}

public class ProtocolMessage
{
    public MessageType Type { get; set; }
    public string PayloadJson { get; set; } = "";
}

public class PairRequestPayload
{
    public string DeviceId { get; set; } = "";
    public string DisplayName { get; set; } = "";
    public string PublicKeyBase64 { get; set; } = ""; // ECDH public key
    public string Pin { get; set; } = ""; // Optional for PIN pairing
}

public class PairResponsePayload
{
    public bool Accepted { get; set; }
    public string PublicKeyBase64 { get; set; } = "";
}

public class TransferOfferPayload
{
    public string TransferId { get; set; } = "";
    public string FileName { get; set; } = "";
    public long TotalBytes { get; set; }
    public bool IsFolder { get; set; }
    public int ItemCount { get; set; }
    public string Hash { get; set; } = "";
}

public class TransferAcceptPayload
{
    public string TransferId { get; set; } = "";
    public bool Accepted { get; set; }
}

public class FileChunkPayload
{
    public string TransferId { get; set; } = "";
    public string RelativePath { get; set; } = ""; // Used for folders
    public long Offset { get; set; }
    // Binary data follows the JSON message in the stream for this type
}

public static class ProtocolFramer
{
    // Write length-prefixed message
    public static async Task WriteMessageAsync(Stream stream, ProtocolMessage message, ReadOnlyMemory<byte> binaryPayload, CancellationToken token)
    {
        var json = JsonSerializer.Serialize(message);
        var jsonBytes = Encoding.UTF8.GetBytes(json);
        
        // Format: [4 bytes JSON length] [JSON bytes] [4 bytes binary length] [binary bytes]
        var header = new byte[8];
        BitConverter.TryWriteBytes(header.AsSpan(0, 4), jsonBytes.Length);
        BitConverter.TryWriteBytes(header.AsSpan(4, 4), binaryPayload.Length);
        
        await stream.WriteAsync(header, token);
        await stream.WriteAsync(jsonBytes, token);
        if (!binaryPayload.IsEmpty)
        {
            await stream.WriteAsync(binaryPayload, token);
        }
    }

    public static async Task<(ProtocolMessage Message, byte[]? BinaryPayload, int BinLen)> ReadMessageAsync(Stream stream, CancellationToken token)
    {
        var header = new byte[8];
        if (!await TryReadExactAsync(stream, header, token))
            throw new EndOfStreamException();

        var jsonLen = BitConverter.ToInt32(header, 0);
        var binLen = BitConverter.ToInt32(header, 4);

        if (jsonLen < 0 || binLen < 0 || jsonLen > 1024 * 1024 || binLen > 1024 * 1024 * 5) // Enforce strict limits
            throw new InvalidDataException("Message length invalid");

        var jsonBytes = new byte[jsonLen];
        if (!await TryReadExactAsync(stream, jsonBytes, token))
            throw new EndOfStreamException();

        var json = Encoding.UTF8.GetString(jsonBytes);
        var message = JsonSerializer.Deserialize<ProtocolMessage>(json) 
            ?? throw new InvalidDataException("Invalid JSON message");

        byte[]? binaryPayload = null;
        if (binLen > 0)
        {
            binaryPayload = System.Buffers.ArrayPool<byte>.Shared.Rent(binLen);
            try
            {
                if (!await TryReadExactAsync(stream, binaryPayload.AsMemory(0, binLen), token))
                    throw new EndOfStreamException();
            }
            catch
            {
                System.Buffers.ArrayPool<byte>.Shared.Return(binaryPayload);
                throw;
            }
        }

        return (message, binaryPayload, binLen);
    }

    private static async Task<bool> TryReadExactAsync(Stream stream, Memory<byte> buffer, CancellationToken token)
    {
        int totalRead = 0;
        while (totalRead < buffer.Length)
        {
            int read = await stream.ReadAsync(buffer.Slice(totalRead), token);
            if (read == 0) return false;
            totalRead += read;
        }
        return true;
    }

    private static async Task<bool> TryReadExactAsync(Stream stream, byte[] buffer, CancellationToken token)
    {
        int totalRead = 0;
        while (totalRead < buffer.Length)
        {
            int read = await stream.ReadAsync(buffer.AsMemory(totalRead, buffer.Length - totalRead), token);
            if (read == 0) return false;
            totalRead += read;
        }
        return true;
    }
}

File: app/src/main/java/com/ioristudios/crossdroid/backend/network/CrossDroidProtocol.kt
package com.ioristudios.crossdroid.backend.network

import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

enum class MessageType {
    PairRequest,
    PairResponse,
    TransferOffer,
    TransferAccept,
    TransferReject,
    TransferCancel,
    FileChunk
}

data class ProtocolMessage(
    var Type: MessageType = MessageType.PairRequest,
    var PayloadJson: String = ""
)

data class PairRequestPayload(
    var DeviceId: String = "",
    var DisplayName: String = "",
    var PublicKeyBase64: String = "",
    var Pin: String = ""
)

data class PairResponsePayload(
    var Accepted: Boolean = false,
    var PublicKeyBase64: String = ""
)

data class TransferOfferPayload(
    var TransferId: String = "",
    var FileName: String = "",
    var TotalBytes: Long = 0,
    var IsFolder: Boolean = false,
    var ItemCount: Int = 0,
    var Hash: String = "",
    var Offset: Long = 0
)

data class TransferAcceptPayload(
    var TransferId: String = "",
    var Accepted: Boolean = false,
    var RequestedOffset: Long = 0
)

data class FileChunkPayload(
    var TransferId: String = "",
    var RelativePath: String = "",
    var Offset: Long = 0
)

object ProtocolFramer {
    private val gson = Gson()

    suspend fun writeMessageAsync(
        stream: OutputStream,
        message: ProtocolMessage,
        binaryPayload: ByteArray?,
        binLenOverride: Int = -1
    ) = withContext(Dispatchers.IO) {
        val json = gson.toJson(message)
        val jsonBytes = json.toByteArray(Charsets.UTF_8)
        val binLen = if (binLenOverride >= 0) binLenOverride else (binaryPayload?.size ?: 0)

        val header = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN)
        header.putInt(jsonBytes.size)
        header.putInt(binLen)

        stream.write(header.array())
        stream.write(jsonBytes)
        if (binaryPayload != null && binLen > 0) {
            stream.write(binaryPayload, 0, binLen)
        }
        stream.flush()
    }

    suspend fun readMessageAsync(
        stream: InputStream
    ): Triple<ProtocolMessage, ByteArray?, Int> = withContext(Dispatchers.IO) {
        val header = ByteArray(8)
        if (!tryReadExact(stream, header)) throw EOFException()

        val headerBuffer = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN)
        val jsonLen = headerBuffer.int
        val binLen = headerBuffer.int

        if (jsonLen < 0 || binLen < 0 || jsonLen > 1024 * 1024 || binLen > 1024 * 1024 * 5) {
            throw java.io.IOException("Message length invalid")
        }

        val jsonBytes = ByteArray(jsonLen)
        if (!tryReadExact(stream, jsonBytes)) throw EOFException()

        val json = String(jsonBytes, Charsets.UTF_8)
        val message = gson.fromJson(json, ProtocolMessage::class.java)
            ?: throw java.io.IOException("Invalid JSON message")

        var binaryPayload: ByteArray? = null
        if (binLen > 0) {
            binaryPayload = ByteArray(binLen)
            if (!tryReadExact(stream, binaryPayload)) throw EOFException()
        }

        Triple(message, binaryPayload, binLen)
    }

    private fun tryReadExact(stream: InputStream, buffer: ByteArray): Boolean {
        var totalRead = 0
        while (totalRead < buffer.size) {
            val read = stream.read(buffer, totalRead, buffer.size - totalRead)
            if (read == -1) return false
            totalRead += read
        }
        return true
    }
}
